README.md
## Pooled SQLite Client for Node.js Apps

<!-- [![NPM version](http://img.shields.io/npm/v/sqlite.svg?style=flat-square)](https://www.npmjs.com/package/sqlite) -->
<!-- [![NPM downloads](http://img.shields.io/npm/dm/sqlite.svg?style=flat-square)](https://www.npmjs.com/package/sqlite) -->
<!-- [![Build Status](http://img.shields.io/travis/kriasoft/node-sqlite/master.svg?style=flat-square)](https://travis-ci.org/kriasoft/node-sqlite) -->
<!-- [![Dependency Status](http://img.shields.io/david/kriasoft/node-sqlite.svg?style=flat-square)](https://david-dm.org/kriasoft/node-sqlite) -->

> A wrapper library that adds ES6 promises, an SQL-based migrations API, connection pooling, and managed transactions to [sqlite3](https://github.com/mapbox/node-sqlite3/) ([docs](https://github.com/mapbox/node-sqlite3/wiki)). Originally based on the [sqlite](https://github.com/kriasoft/node-sqlite/) library.

---


### How to Install

```sh
$ npm install sqlite-pool --save
```


### How to Use

This module has a similar API as the original `sqlite3` library ([docs](https://github.com/mapbox/node-sqlite3/wiki/API)), except that its equivalent API methods return ES6 Promises and do not accept callback arguments. Full API listings can be found in [API.md](https://github.com/rneilson/node-sqlite-pool/tree/master/API.md).

Below is an example of how to use it with [Node.js](https://nodejs.org) and [Express](http://expressjs.com/starter/hello-world.html), using [Bluebird](http://bluebirdjs.com/) instead of native Promises:

```js
const express = require('express');
const Promise = require('bluebird');
const Sqlite = require('sqlite-pool');

const app = express();
const port = process.env.PORT || 3000;
const db = new Sqlite('./database.sqlite', { Promise });

app.get('/post/:id', (req, res, next) => {
  return Promise.all([
    db.get('SELECT * FROM Post WHERE id = ?', req.params.id),
    db.all('SELECT * FROM Category')
  ]).then(([post, categories]) => {
    res.render('post', { post, categories });
  }).catch(next);
});

// Launch Node.js app
app.listen(port);
```

**NOTE**: For Node.js v5 and below use `var Sqlite = require('sqlite-pool/legacy');`.

### Multiple Connections

Due to the asynchronous interface which the [sqlite3](https://github.com/mapbox/node-sqlite3/) Node.js library provides, isolation and query ordering is not guaranteed within any given connection. This module uses the [generic-pool](https://github.com/coopernurse/node-pool) library to create multiple connections to an SQLite database if desired, and queuing requests made using methods of the `Sqlite` class.

The minimum/maximum number of open connections can be set with the `min` and `max` options when calling `new Sqlite()` (see the [API reference](https://github.com/rneilson/node-sqlite-pool/tree/master/API.md) for the full list of options);

Below is an example with the `Sqlite.use()` method of how one connection may perform a transaction isolated from other calls to the same `Sqlite` instance:

```javascript
const express = require('express');
const Promise = require('bluebird');
const Sqlite = require('sqlite-pool');

const app = express();
const port = process.env.PORT || 3000;
const db = new Sqlite('./database.sqlite', { Promise, min: 2, max: 4 });

app.get('/post/:id', (req, res, next) => {
  return Promise.all([
    // This will acquire a database connection, and release it
    // once the returned Promise is resolved or rejected
    db.use((conn) => {
      let id = req.params.id;
      // This Promise chain will begin a transaction, and either
      // commit if successful or rollback if an error is thrown
      return conn.exec('BEGIN IMMEDIATE')
        .then(() => conn.get('SELECT * FROM Post WHERE id = ?', id))
        .then((post) => {
          if (post === undefined) {
            throw new Error(`Post id ${id} not found`);
          }
          return conn.run('UPDATE Post SET views = views + 1 WHERE id = ?', id)
            .then(() => conn.exec('COMMIT'))
            .then(() => post);
        })
        .catch(err => conn.exec('ROLLBACK').then(() => Promise.reject(err)));
    }),
    // This query will run in a separate connection, outside of
    // the above transaction
    db.all('SELECT * FROM Category')
  ]).then(([post, categories]) => {
    res.render('post', { post, categories });
  }).catch(next);
});

// Launch Node.js app
app.listen(port);
```

### Transactions

TODO

### Migrations

This module comes with a lightweight migrations API that works with [SQL-based migration files](https://github.com/rneilson/node-sqlite-pool/tree/master/migrations)
as the following example demonstrates:

##### `migrations/001-initial-schema.sql`

```sql
-- Up

CREATE TABLE Category (
  id   INTEGER PRIMARY KEY,
  name TEXT    NOT NULL
);

CREATE TABLE Post (
  id          INTEGER PRIMARY KEY,
  categoryId  INTEGER NOT NULL,
  title       TEXT    NOT NULL,
  views       NUMERIC NOT NULL DEFAULT 0,
  isPublished NUMERIC NOT NULL DEFAULT 0,
  CONSTRAINT Post_fk_categoryId FOREIGN KEY (categoryId)
    REFERENCES Category (id) ON UPDATE CASCADE ON DELETE CASCADE,
  CONSTRAINT Post_ck_isPublished CHECK (isPublished IN (0, 1))
);

INSERT INTO Category (id, name) VALUES (1, 'Business');
INSERT INTO Category (id, name) VALUES (2, 'Technology');

-- Down

DROP TABLE Post;
DROP TABLE Category;
```

##### `migrations/002-missing-index.sql`

```sql
-- Up
CREATE INDEX Post_ix_categoryId ON Post (categoryId);

-- Down
DROP INDEX Post_ix_categoryId;
```

##### `app.js` (Node.js/Express)

```js
const express = require('express');
const Promise = require('bluebird');
const Sqlite = require('sqlite-pool');

const app = express();
const port = process.env.PORT || 3000;
const db = new Sqlite('./database.sqlite', { Promise });

app.use(/* app routes */);

Promise.resolve()
  // First, try to update the database schema to the latest version
  .then(() => db.migrate({
    force: 'last',                  // Default: false
    table: 'migrations',            // Default: 'migrations'
    migrationsPath: './migrations'  // Default: './migrations'
  }))
  .catch(err => console.error(err.stack))
  // Finally, launch Node.js app
  .finally(() => app.listen(port));
```

**NOTE**: For the development environment, while working on the database schema, you may want to set `force: 'last'` (default `false`) that will force the migration API to rollback and re-apply the latest migration over again each time when Node.js app launches. 


### References

* [SQLite Documentation](https://www.sqlite.org/docs.html), e.g. [SQL Syntax](https://www.sqlite.org/lang.html), [Data Types](https://www.sqlite.org/datatype3.html) etc. on SQLite.org


### License

The MIT License © 2017 Raymond Neilson. All rights reserved.
Original work © 2015-2016 Kriasoft, LLC. All rights reserved.

---
Original library by Konstantin Tarkus ([@koistya](https://twitter.com/koistya)) and [contributors](https://github.com/rneilson/node-sqlite-pool/graphs/contributors)


CHANGELOG.md
## Pooled SQLite Client Change Log

All notable changes to this project will be documented in this file.

### [v1.0.1]
> 2017-01-31

- Fix for migration sorting error (see [node-sqlite issue #25](https://github.com/kriasoft/node-sqlite/pull/25))
- Minor changes to satisfy linting.

### [v1.0.0]
> 2017-01-30

- Initial release, based on the [node-sqlite](https://github.com/kriasoft/node-sqlite/) library for wrapping [node-sqlite3](https://github.com/mapbox/node-sqlite3), and incorporating the [generic-pool](https://github.com/coopernurse/node-pool) library to manage multiple connections to a database.

[unreleased]: https://github.com/rneilson/node-sqlite-pool/compare/v1.0.0...HEAD
[v1.0.0]: https://github.com/rneilson/node-sqlite-pool/compare/v0.0.0...v1.0.0
[v1.0.1]: https://github.com/rneilson/node-sqlite-pool/compare/v1.0.0...v1.0.1


test/spec.js
/**
 * Pooled SQLite client library for Node.js
 * Based on the node-sqlite library
 *
 * Copyright © 2017 Raymond Neilson. All rights reserved.
 *
 * Original work copyright © 2016 Kriasoft, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

const Sqlite = require('../build/main');
const expect = require('chai').expect;

it('Should open a database connection', (done) => {
  const db = new Sqlite(':memory:');
  let p = Promise.resolve();
  p = p.then(() => db.exec('CREATE TABLE tbl (col TEXT)'));
  p = p.then(() => db.exec('INSERT INTO tbl VALUES ("test")'));
  p = p.then(() => db.get('SELECT col FROM tbl').then((result) => {
    expect(result).to.be.deep.equal({ col: 'test' });
  }));
  p = p.then(() => db.all('SELECT col FROM tbl').then((result) => {
    expect(result).to.be.deep.equal([{ col: 'test' }]);
  }));
  p = p.then(() => db.all('SELECT * FROM tbl WHERE col = ?', 'test').then((result) => {
    expect(result).to.have.length(1);
  }));
  p = p.then(() => db.run('UPDATE tbl SET col = ? WHERE col = ?', 'foo', 'test')).then((stmt) => {
    // Cannot use deep equals because stmt is a Statement instance
    expect(stmt.lastID).to.equal(1);
    expect(stmt.changes).to.equal(1);
    expect(stmt.sql).to.equal('UPDATE tbl SET col = ? WHERE col = ?');
  });
  p = p.then(() => db.get('SELECT col FROM tbl').then((result) => {
    expect(result).to.be.deep.equal({ col: 'foo' });
  }));
  p = p.then(() => db.close());
  p.then(done, done);
});

it('Should allow named parameters to be used', (done) => {
  const db = new Sqlite(':memory:');
  let p = Promise.resolve();
  p = p.then(() => db.exec('CREATE TABLE tbl (col TEXT)'));
  p = p.then(() => db.run('INSERT INTO tbl(col) VALUES (:col)', { ':col': 'something' })
    .then((stmt) => {
      expect(stmt.lastID).to.equal(1);
    }));
  p = p.then(() => db.get('SELECT col FROM tbl WHERE 1 = ? AND 5 = ?5', { 1: 1, 5: 5 })
    .then((result) => {
      expect(result).to.be.deep.equal({ col: 'something' });
    }));
  p = p.then(() => db.run('INSERT INTO tbl(col) VALUES ($col)', { $col: 'other thing' })
    .then((stmt) => {
      expect(stmt.lastID).to.equal(2);
    }));
  p = p.then(() => db.each('SELECT col FROM tbl WHERE ROWID = ?', [2], (result) => {
    expect(result).to.be.deep.equal({ col: 'other thing' });
  }).then((rowsCount) => {
    expect(rowsCount).to.equal(1);
  }));
  p = p.then(() => db.all('SELECT col FROM tbl WHERE 13 = @thirteen ORDER BY col DESC', {
    '@thirteen': 13,
  }).then((results) => {
    expect(results).to.be.deep.equal([{ col: 'something' }, { col: 'other thing' }]);
  }));
  p = p.then(() => db.close());
  p.then(done, done);
});

it('Should allow named parameters to be used with prepared statements', (done) => {
  const db = new Sqlite(':memory:');
  let p = db.use((conn) => {
    let q = Promise.resolve();
    q = q.then(() => conn.exec('CREATE TABLE tbl (col TEXT)'));
    q = q.then(() => conn.prepare('INSERT INTO tbl(col) VALUES (:col)', { ':col': 'some text' })
      .then(stmt => stmt.run())
      .then((stmt) => {
        expect(stmt.lastID).to.equal(1);
        return stmt.finalize();
      }));
    q = q.then(() => conn.prepare('SELECT col FROM tbl WHERE 1 = ? AND 5 = ?5')
      .then(stmt => stmt.bind({ 1: 1, 5: 5 }))
      .then(stmt => stmt.get()
        .then((result) => {
          expect(result).to.be.deep.equal({ col: 'some text' });
          return stmt.finalize();
        })));
    q = q.then(() => conn.prepare('INSERT INTO tbl(col) VALUES ($col)')
      .then(stmt => stmt.run({ $col: 'other text' }))
      .then((stmt) => {
        expect(stmt.lastID).to.equal(2);
        return stmt.finalize();
      }));
    q = q.then(() => conn.prepare('SELECT col FROM tbl WHERE ROWID = ?')
      .then(stmt => stmt.each([2], (result) => {
        expect(result).to.be.deep.equal({ col: 'other text' });
      }).then((rowsCount) => {
        expect(rowsCount).to.equal(1);
        return stmt.finalize();
      })));
    q = q.then(() => conn.prepare('SELECT col FROM tbl WHERE 13 = @thirteen ORDER BY col DESC')
      .then(stmt => stmt.all({ '@thirteen': 13 })
        .then((results) => {
          expect(results).to.be.deep.equal([{ col: 'some text' }, { col: 'other text' }]);
          return stmt.finalize();
        })));
    return q;
  });
  p = p.then(() => db.close());
  p.then(done, done);
});

it('Should allow chaining Statement.run() calls', (done) => {
  const db = new Sqlite(':memory:');
  let p = db.use((conn) => {
    let q = Promise.resolve();
    q = q.then(() => conn.exec('CREATE TABLE tbl (col1 TEXT, col2 TEXT, col3 TEXT)'));
    q = q.then(() => conn.prepare('INSERT INTO tbl(col1, col2, col3) VALUES (?, ?, ?)')
      .then(stmt => stmt.run('a1', 'a2', 'a3')).then((stmt) => {
        expect(stmt.lastID).to.equal(1);
        return stmt.run('b1', 'b2', 'b3');
      })
      .then((stmt) => {
        expect(stmt.lastID).to.equal(2);
        return stmt.finalize();
      }));
    q = q.then(() => conn.all('SELECT col1, col2, col3 FROM tbl').then((results) => {
      expect(results).to.be.deep.equal([{
        col1: 'a1',
        col2: 'a2',
        col3: 'a3',
      }, {
        col1: 'b1',
        col2: 'b2',
        col3: 'b3',
      }]);
    }));
    return q;
  });
  p = p.then(() => db.close());
  p.then(done, done);
});

it('Should handle BLOBs', (done) => {
  const buf = Buffer.from('SGVsbG8gd29ybGQh', 'base64');
  const db = new Sqlite(':memory:');
  let p = Promise.resolve();
  p = p.then(() => db.exec('CREATE TABLE dat (b BLOB)'));
  p = p.then(() => db.run('INSERT INTO dat(b) VALUES(?)', buf).then((stmt) => {
    expect(stmt.lastID).to.equal(1);
  }));
  p = p.then(() => db.get('SELECT b FROM dat').then((result) => {
    expect(result.b).to.be.instanceof(Buffer);
    expect(result.b.toString('utf8')).to.equal('Hello world!');
  }));
  p.then(done, done);
});

it('Should migrate the database', (done) => {
  const db = new Sqlite(':memory:');
  let p = Promise.resolve();
  p = p.then(() => db.migrate());
  p = p.then(() => db.all('SELECT id, name FROM migrations').then((result) => {
    expect(result).to.be.deep.equal([{ id: 1, name: 'initial' }, { id: 2, name: 'some-feature' }]);
  }));
  p = p.then(() => db.all('SELECT * FROM Category').then((result) => {
    expect(result).to.be.deep.equal([{ id: 1, name: 'Test' }]);
  }));
  p = p.then(() => db.close());
  p.then(done, done);
});


src/main.js
/**
 * Pooled SQLite client library for Node.js
 * Based on the node-sqlite library
 *
 * Copyright © 2017 Raymond Neilson. All rights reserved.
 *
 * Original work copyright © 2016 Kriasoft, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import fs from 'fs';
import path from 'path';
import sqlite3 from 'sqlite3';
import genericPool from 'generic-pool';
import Database from './Database';
import { isThenable, asyncRunner } from './utils';

// Default options
const defaults = {
  // sqlite defaults
  mode: null,
  busyTimeout: 1000,
  foreignKeys: true,
  walMode: true,
  loadExtensions: [],

  // pool defaults
  min: 1,
  max: 4,
  acquireTimeout: 1000,

  // internal defaults
  trxImmediate: true,
  delayRelease: true,

  // general defaults
  Promise: global.Promise,
};

class Sqlite {
  constructor (filename, options) {
    // Extract options
    const {
      mode,
      busyTimeout,
      foreignKeys,
      walMode,
      loadExtensions,
      min,
      max,
      trxImmediate,
      delayRelease,
      acquireTimeout,
      Promise,
    } = Object.assign({}, defaults, options);

    // Re-consolidate options
    this._pool_opts = { min, max, Promise, acquireTimeoutMillis: acquireTimeout };
    this._sqlite_opts = { filename, mode, busyTimeout, foreignKeys, walMode };
    this._sqlite_ext = loadExtensions;
    this.trxImmediate = trxImmediate;
    this.delayRelease = delayRelease;
    this.Promise = Promise;

    // Async runner
    this.async = asyncRunner(Promise);

    // Special case min/max for anonymous or in-memory database
    if (filename === '' || filename === ':memory:') {
      this._pool_opts.min = 1;
      this._pool_opts.max = 1;
    }

    // Factory functions for generic-pool
    this._pool_factory = {
      create: () => this._create(),

      destroy: connection => this._destroy(connection),
    };

    // Create pool
    this._pool = genericPool.createPool(this._pool_factory, this._pool_opts);
  }

  _create () {
    return this.async(function* _createAsync () {
      const Promise = this.Promise;
      const trxImmediate = this.trxImmediate;
      const options = this._sqlite_opts;

      // Create database connection, wait until open complete
      const connection = yield new Promise((resolve, reject) => {
        let driver;
        const callback = (err) => {
          if (err) {
            return reject(err);
          }
          return resolve(new Database(driver, { Promise, trxImmediate }));
        };

        if (options.mode !== null) {
          driver = new sqlite3.Database(options.filename, options.mode, callback);
        }
        else {
          driver = new sqlite3.Database(options.filename, callback);
        }

        // Busy timeout default hardcoded to 1000ms, so
        // only configure if a different value given
        if (options.busyTimeout !== 1000) {
          driver.configure('busyTimeout', options.busyTimeout);
        }
      });

      // Load extensions
      // Await each for consistency
      for (const extension of this._sqlite_ext) {
        const extensionPath = path.resolve(extension);
        yield new Promise((resolve, reject) => {
          connection.driver.loadExtension(extensionPath, (err) => {
            if (err) {
              return reject(err);
            }
            return resolve();
          });
        });
      }

      // Set foreign keys and/or WAL mode as appropriate
      if (options.foreignKeys) {
        yield connection.exec('PRAGMA foreign_keys = ON;');
      }
      if (options.walMode) {
        yield connection.exec('PRAGMA journal_mode = WAL;');
      }

      // Return now-configured db connection
      return connection;
    });
  }

  _destroy (connection) {
    return new this.Promise((resolve, reject) => {
      connection.driver.close((err) => {
        if (err) {
          return reject(err);
        }
        return resolve();
      });
    });
  }

  _release (connection) {
    if (this.delayRelease) {
      return setImmediate(() => this._pool.release(connection));
    }
    return this._pool.release(connection);
  }

  _acquireRelease (fn, async = false) {
    return this.async(function* _acquireReleaseAsync () {
      const connection = yield this._pool.acquire();
      let result;
      try {
        if (async) {
          // Run fn as async (generator)
          result = yield this.async(fn, connection);
        }
        else {
          // Pass connection to function
          result = yield fn.call(this, connection);
        }
      }
      finally {
        this._release(connection);
      }
      return result;
    });
  }

  close () {
    return this.async(function* _closeAsync () {
      yield this._pool.drain();
      yield this._pool.clear();
    });
  }

  exec (...args) {
    return this._acquireRelease(conn => conn.exec(...args)).then(() => {});
  }

  run (...args) {
    return this._acquireRelease(conn => conn.run(...args));
  }

  get (...args) {
    return this._acquireRelease(conn => conn.get(...args));
  }

  all (...args) {
    return this._acquireRelease(conn => conn.all(...args));
  }

  each (...args) {
    return this._acquireRelease(conn => conn.each(...args));
  }

  use (fn) {
    return this._acquireRelease((conn) => {
      // Pass connection to function
      const result = fn.call(this, conn);

      // If function didn't return a thenable, wait
      return isThenable(result) ? result : conn.wait().then(() => result);
    });
  }

  useAsync (gen) {
    return this._acquireRelease(gen, true);
  }

  transaction (fn, immediate = this.trxImmediate) {
    return this._acquireRelease(conn => conn.transaction(fn, immediate));
  }

  transactionAsync (gen, immediate = this.trxImmediate) {
    return this._acquireRelease(conn => conn.transactionAsync(gen, immediate));
  }

  /**
   * Migrates database schema to the latest version
   */
  migrate ({ force, table = 'migrations', migrationsPath = './migrations' } = {}) {
    return this.async(function* _migrateAsync () {
      const Promise = this.Promise;
      const location = path.resolve(migrationsPath);

      // Get the list of migration files, for example:
      //   { id: 1, name: 'initial', filename: '001-initial.sql' }
      //   { id: 2, name: 'feature', fielname: '002-feature.sql' }
      const migrations = yield new Promise((resolve, reject) => {
        fs.readdir(location, (err, files) => {
          if (err) {
            reject(err);
          }
          else {
            resolve(files
              .map(x => x.match(/^(\d+).(.*?)\.sql$/))
              .filter(x => x !== null)
              .map(x => ({ id: Number(x[1]), name: x[2], filename: x[0] }))
              .sort((a, b) => a.id > b.id));
          }
        });
      });

      if (!migrations.length) {
        throw new Error(`No migration files found in '${location}'.`);
      }

      // Get the list of migrations, for example:
      // { id: 1, name: 'initial', filename: '001-initial.sql', up: ..., down: ... }
      // { id: 2, name: 'feature', fielname: '002-feature.sql', up: ..., down: ... }
      yield Promise.all(migrations.map(migration => new Promise((resolve, reject) => {
        const filename = path.join(location, migration.filename);
        fs.readFile(filename, 'utf-8', (err, data) => {
          if (err) {
            reject(err);
          }
          else {
            const [up, down] = data.split(/^--\s+?down/mi);
            if (!down) {
              reject(new Error(
                `The file ${migration.filename} is missing a '-- Down' separator.`
              ));
            }
            else {
              // Remove comments and trim whitespaces
              /* eslint-disable no-param-reassign */
              migration.up = up.replace(/^--.*?$/gm, '').trim();
              migration.down = down.replace(/^--.*?$/gm, '').trim();
              /* eslint-enable no-param-reassign */
              resolve();
            }
          }
        });
      })));

      yield this.useAsync(function* _runMigrationsAsync (conn) {
        // Create a database table for migrations meta data if it doesn't exist
        yield conn.run(`CREATE TABLE IF NOT EXISTS "${table}" (
    id   INTEGER PRIMARY KEY,
    name TEXT    NOT NULL,
    up   TEXT    NOT NULL,
    down TEXT    NOT NULL
  )`);

        // Get the list of already applied migrations
        let dbMigrations = yield conn.all(
          `SELECT id, name, up, down FROM "${table}" ORDER BY id ASC`,
        );

        // Undo migrations that exist only in the database but not in files,
        // also undo the last migration if the `force` option was set to `last`.
        const lastMigration = migrations[migrations.length - 1];
        const prevMigrations = dbMigrations.slice().sort((a, b) => a.id < b.id);
        for (const migration of prevMigrations) {
          if (!migrations.some(x => x.id === migration.id) ||
              (force === 'last' && migration.id === lastMigration.id)) {
            yield conn.transactionAsync(function* _downAsync (trx) {
              yield trx.exec(migration.down);
              yield trx.run(`DELETE FROM "${table}" WHERE id = ?`, migration.id);
            });
            dbMigrations = dbMigrations.filter(x => x.id !== migration.id);
          }
          else {
            break;
          }
        }

        // Apply pending migrations
        const lastMigrationId = dbMigrations.length
                              ? dbMigrations[dbMigrations.length - 1].id
                              : 0;
        for (const migration of migrations) {
          if (migration.id > lastMigrationId) {
            yield conn.transactionAsync(function* _upAsync (trx) {
              yield trx.exec(migration.up);
              yield trx.run(
                `INSERT INTO "${table}" (id, name, up, down) VALUES (?, ?, ?, ?)`,
                migration.id, migration.name, migration.up, migration.down
              );
            });
          }
        }
      });

      return this;
    });
  }
}

export default Sqlite;


src/Sqlite.js
/**
 * Pooled SQLite client library for Node.js
 * Based on the node-sqlite library
 *
 * Copyright © 2017 Raymond Neilson. All rights reserved.
 *
 * Original work copyright © 2016 Kriasoft, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import fs from 'fs';
import path from 'path';
import sqlite3 from 'sqlite3';
import genericPool from 'generic-pool';
import Database from './Database';
import { isThenable, asyncRunner } from './utils';

// Default options
const defaults = {
  // sqlite defaults
  mode: null,
  verbose: false,
  busyTimeout: 1000,
  foreignKeys: true,
  walMode: true,
  loadExtensions: [],

  // pool defaults
  min: 1,
  max: 4,
  acquireTimeout: 1000,

  // internal defaults
  trxImmediate: true,
  delayRelease: true,

  // general defaults
  Promise: global.Promise,
};

class Sqlite {
  constructor (filename = ':memory:', options = {}) {
    // Extract options
    const {
      mode,
      verbose,
      busyTimeout,
      foreignKeys,
      walMode,
      loadExtensions,
      min,
      max,
      trxImmediate,
      delayRelease,
      acquireTimeout,
      Promise,
    } = Object.assign({}, defaults, options);

    // Re-consolidate options
    this._pool_opts = { min, max, Promise, acquireTimeoutMillis: acquireTimeout };
    this._sqlite_opts = { mode, verbose, busyTimeout, foreignKeys, walMode };
    this._sqlite_file = filename;
    this._sqlite_extn = loadExtensions;
    this._immediate = trxImmediate;
    this.delayRelease = delayRelease;
    this.Promise = Promise;

    // Async runner
    this._async = asyncRunner(Promise);

    // Special case min/max for anonymous or in-memory database
    if (filename === '' || filename === ':memory:') {
      this._pool_opts.min = 1;
      this._pool_opts.max = 1;
    }

    // Factory functions for generic-pool
    this._pool_factory = {
      create: () => this._create(),

      destroy: connection => this._destroy(connection),
    };

    // Create pool
    this._pool = genericPool.createPool(this._pool_factory, this._pool_opts);
  }

  static get OPEN_READONLY () {
    return sqlite3.OPEN_READONLY;
  }

  static get OPEN_READWRITE () {
    return sqlite3.OPEN_READWRITE;
  }

  static get OPEN_CREATE () {
    return sqlite3.OPEN_CREATE;
  }

  _create () {
    return this._async(function* _createAsync () {
      const Promise = this.Promise;
      const trxImmediate = this._immediate;
      const options = this._sqlite_opts;
      const filename = this._sqlite_file;

      // Create database connection, wait until open complete
      const connection = yield new Promise((resolve, reject) => {
        let driver;
        const callback = (err) => {
          if (err) {
            return reject(err);
          }
          return resolve(new Database(driver, { Promise, trxImmediate }));
        };

        if (options.mode !== null) {
          driver = new sqlite3.Database(filename, options.mode, callback);
        }
        else {
          driver = new sqlite3.Database(filename, callback);
        }

        // Can't reset this per connection
        if (options.verbose) {
          driver.verbose();
        }

        // Busy timeout default hardcoded to 1000ms, so
        // only configure if a different value given
        if (options.busyTimeout !== 1000) {
          driver.configure('busyTimeout', options.busyTimeout);
        }
      });

      // Await each for consistency
      // Load extensions
      for (const extension of this._sqlite_extn) {
        const extensionPath = path.resolve(extension);
        yield new Promise((resolve, reject) => {
          connection.driver.loadExtension(extensionPath, (err) => {
            if (err) {
              return reject(err);
            }
            return resolve();
          });
        });
      }

      // Set foreign keys and/or WAL mode as appropriate
      if (options.foreignKeys) {
        yield connection.exec('PRAGMA foreign_keys = ON;');
      }
      if (options.walMode) {
        yield connection.exec('PRAGMA journal_mode = WAL;');
      }

      // Return now-configured db connection
      return connection;
    });
  }

  _destroy (connection) {
    return new this.Promise((resolve, reject) => {
      connection.driver.close((err) => {
        if (err) {
          return reject(err);
        }
        return resolve();
      });
    });
  }

  _release (connection) {
    if (this.delayRelease) {
      return setImmediate(() => this._pool.release(connection));
    }
    return this._pool.release(connection);
  }

  _acquireRelease (fn, isAsync = false) {
    return this._async(function* _acquireReleaseAsync () {
      const connection = yield this._pool.acquire();
      let result;
      try {
        if (isAsync) {
          // Run fn as async (generator)
          result = yield this._async(fn, connection);
        }
        else {
          // Pass connection to function
          result = yield fn.call(this, connection);
        }
      }
      finally {
        this._release(connection);
      }
      return result;
    });
  }

  close () {
    return this._async(function* _closeAsync () {
      yield this._pool.drain();
      yield this._pool.clear();
    });
  }

  exec (...args) {
    return this._acquireRelease(conn => conn.exec(...args)).then(() => {});
  }

  run (...args) {
    return this._acquireRelease(conn => conn.run(...args));
  }

  get (...args) {
    return this._acquireRelease(conn => conn.get(...args));
  }

  all (...args) {
    return this._acquireRelease(conn => conn.all(...args));
  }

  each (...args) {
    return this._acquireRelease(conn => conn.each(...args));
  }

  use (fn) {
    return this._acquireRelease((conn) => {
      // Pass connection to function
      const result = fn.call(this, conn);

      // If function didn't return a thenable, wait
      return isThenable(result) ? result : conn.wait().then(() => result);
    });
  }

  useAsync (gen) {
    return this._acquireRelease(gen, true);
  }

  transaction (fn, immediate = this._immediate) {
    return this._acquireRelease(conn => conn.transaction(fn, immediate));
  }

  transactionAsync (gen, immediate = this._immediate) {
    return this._acquireRelease(conn => conn.transactionAsync(gen, immediate));
  }

  /**
   * Migrates database schema to the latest version
   */
  migrate ({ force, table = 'migrations', migrationsPath = './migrations' } = {}) {
    return this._async(function* _migrateAsync () {
      const Promise = this.Promise;
      const location = path.resolve(migrationsPath);

      // Get the list of migration files, for example:
      //   { id: 1, name: 'initial', filename: '001-initial.sql' }
      //   { id: 2, name: 'feature', fielname: '002-feature.sql' }
      const migrations = yield new Promise((resolve, reject) => {
        fs.readdir(location, (err, files) => {
          if (err) {
            reject(err);
          }
          else {
            resolve(files
              .map(x => x.match(/^(\d+).(.*?)\.sql$/))
              .filter(x => x !== null)
              .map(x => ({ id: Number(x[1]), name: x[2], filename: x[0] }))
              .sort((a, b) => (a.id > b.id ? 1 : a.id < b.id ? -1 : 0)));
          }
        });
      });

      if (!migrations.length) {
        throw new Error(`No migration files found in '${location}'.`);
      }

      // Get the list of migrations, for example:
      // { id: 1, name: 'initial', filename: '001-initial.sql', up: ..., down: ... }
      // { id: 2, name: 'feature', fielname: '002-feature.sql', up: ..., down: ... }
      yield Promise.all(migrations.map(migration => new Promise((resolve, reject) => {
        const filename = path.join(location, migration.filename);
        fs.readFile(filename, 'utf-8', (err, data) => {
          if (err) {
            reject(err);
          }
          else {
            const [up, down] = data.split(/^--\s+?down/mi);
            if (!down) {
              reject(new Error(
                `The file ${migration.filename} is missing a '-- Down' separator.`
              ));
            }
            else {
              // Remove comments and trim whitespaces
              /* eslint-disable no-param-reassign */
              migration.up = up.replace(/^--.*?$/gm, '').trim();
              migration.down = down.replace(/^--.*?$/gm, '').trim();
              /* eslint-enable no-param-reassign */
              resolve();
            }
          }
        });
      })));

      yield this.useAsync(function* _runMigrationsAsync (conn) {
        // Create a database table for migrations meta data if it doesn't exist
        yield conn.run(`CREATE TABLE IF NOT EXISTS "${table}" (
    id   INTEGER PRIMARY KEY,
    name TEXT    NOT NULL,
    up   TEXT    NOT NULL,
    down TEXT    NOT NULL
  )`);

        // Get the list of already applied migrations
        let dbMigrations = yield conn.all(
          `SELECT id, name, up, down FROM "${table}" ORDER BY id ASC`,
        );

        // Undo migrations that exist only in the database but not in files,
        // also undo the last migration if the `force` option was set to `last`.
        const lastMigration = migrations[migrations.length - 1];
        const prev = dbMigrations
                      .slice()
                      .sort((a, b) => (a.id < b.id ? 1 : a.id > b.id ? -1 : 0));
        for (const migration of prev) {
          if (!migrations.some(x => x.id === migration.id) ||
              (force === 'last' && migration.id === lastMigration.id)) {
            yield conn.transactionAsync(function* _downAsync (trx) {
              yield trx.exec(migration.down);
              yield trx.run(`DELETE FROM "${table}" WHERE id = ?`, migration.id);
            });
            dbMigrations = dbMigrations.filter(x => x.id !== migration.id);
          }
          else {
            break;
          }
        }

        // Apply pending migrations
        const lastMigrationId = dbMigrations.length
                              ? dbMigrations[dbMigrations.length - 1].id
                              : 0;
        for (const migration of migrations) {
          if (migration.id > lastMigrationId) {
            yield conn.transactionAsync(function* _upAsync (trx) {
              yield trx.exec(migration.up);
              yield trx.run(
                `INSERT INTO "${table}" (id, name, up, down) VALUES (?, ?, ?, ?)`,
                migration.id, migration.name, migration.up, migration.down
              );
            });
          }
        }
      });

      return this;
    });
  }
}

export default Sqlite;
